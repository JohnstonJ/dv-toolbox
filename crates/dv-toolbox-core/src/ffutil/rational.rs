use derive_more::{Deref, DerefMut};
use ffmpeg_next as ffmpeg;
use num::Rational32;

/// Newtype class for converting between num-rational Ratio and FFmpeg AVRational types.  Values
/// will automatically be reduced when making the conversion.
#[derive(Debug, Deref, DerefMut)]
pub struct AVRationalConverter(ffmpeg::Rational);

impl From<AVRationalConverter> for Rational32 {
    fn from(value: AVRationalConverter) -> Self {
        (value.numerator(), value.denominator()).into()
    }
}

impl From<Rational32> for AVRationalConverter {
    fn from(value: Rational32) -> Self {
        Self(value.reduced().into_raw().into())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use googletest::prelude::*;

    #[googletest::test]
    fn test_rational_converter() {
        // ffmpeg::Rational to Rational32 conversion
        expect_that!(Rational32::new(3, 4), eq(AVRationalConverter(ffmpeg::Rational(6, 8)).into()));

        // Rational32 to ffmpeg::Rational conversion
        expect_that!(
            ffmpeg::Rational::new(3, 4),
            eq(*AVRationalConverter::from(Rational32::new_raw(6, 8)))
        );
    }
}
