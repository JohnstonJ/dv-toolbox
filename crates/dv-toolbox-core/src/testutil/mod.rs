use core::str;
use std::{
    collections::{HashMap, HashSet},
    path::PathBuf,
    sync::LazyLock,
};

use data_encoding::{Encoding, HEXUPPER_PERMISSIVE};
use derive_more::{Deref, From};
use googletest::prelude::*;
use regex::Regex;

/// Directory containing test-related data files.
pub(crate) fn test_resource(path: &str) -> PathBuf {
    [env!("CARGO_MANIFEST_DIR"), "resources/test", path].iter().collect()
}

static PADDED_HEX: LazyLock<Encoding> = LazyLock::new(|| {
    let mut spec = HEXUPPER_PERMISSIVE.specification();
    spec.ignore = String::from(" ");
    spec.encoding().unwrap()
});

/// Parse a string like "AA BB CC DD EE" into a byte array
pub(crate) fn from_hex<const N: usize>(value: &str) -> [u8; N] {
    PADDED_HEX
        .decode(value.as_bytes())
        .expect("value should be a valid hexadecimal string")
        .try_into()
        .unwrap()
}

/// Map of test case name to test case data.  See the [`test_case_map`] macro for an example of how
/// to use this.
#[derive(Debug, Deref, From)]
pub(crate) struct TestCaseMap<'a, T>(HashMap<&'a str, T>);

impl<'a, T> TestCaseMap<'a, T> {
    /// Return test case given the test function name generated by rstest
    ///
    /// The parameter needs to be the return value of the [`stdext::function_name`] macro.  The
    /// macro must be called from within a `#[case]` attribute.  Do not call it from within your
    /// test function itself.
    ///
    /// Example input: `my_crate_name::my_module::tests::test_frob::case_1_success_if_foo` which
    /// will return `success`.
    pub(crate) fn get_test_case(&self, test_function_name: &'a str) -> &T {
        let case_name = &GET_TEST_CASE_RE.captures(test_function_name).unwrap()[1];
        assert!(self.contains_key(case_name), "Test case {case_name} is not in test table");
        &self[case_name]
    }
}

/// A lazy-initialized map of test cases.  See the [`test_case_map`] macro for an example of how to
/// use this.
pub(crate) type LazyTestCases<'a, T> = LazyLock<TestCaseMap<'a, T>>;

/// Lazy-initializes a map of [`rstest`] test cases for a given function.  Intended for use as a
/// static variable.
///
/// The advantage of using this over passing structures directly to `#[case]` is that the structures
/// will be auto-formatted with rustfmt.
///
/// # Examples
///
/// ```ignore
/// use rstest::rstest;
/// use stdext::function_name;
///
/// #[derive(Debug)]
/// struct FunctionATestCase {
///     foo: i32,
///     bar: i32,
/// }
///
/// static FUNCTION_A_TEST_CASES: LazyTestCases<FunctionATestCase> = test_case_map!(
///     "first_case",
///     FunctionATestCase { foo: 1, bar: 1 },
///     "second_case",
///     FunctionATestCase { foo: 2, bar: 2 }
/// );
///
/// #[rstest]
/// #[case::first_case(function_name!())]
/// #[case::second_case(function_name!())]
/// fn test_function_a(#[case] test_function_name: &str) {
///     let tc = FUNCTION_A_TEST_CASES.get_test_case(test_function_name);
///
///     assert!(tc.foo > 0);
/// }
///
/// // If you add an entry to FUNCTION_A_TEST_CASES but forget to add a #[case] for it, this will
/// // catch the mistake.
/// test_all_test_cases_ran!(
///     ("test_function_a", &FUNCTION_A_TEST_CASES)
/// );
/// ```
macro_rules! test_case_map {
    ($($test_case_name:expr, $test_case_data:expr),*) => {
        ::std::sync::LazyLock::new(|| ::std::collections::HashMap::<&str, _>::from([
            $(($test_case_name, $test_case_data)),*
        ]).into())
    };
}

pub(crate) use test_case_map;

/// Asserts that all [`rstest`] test cases in the given tables of test cases are actual test cases
/// in this test binary.
///
/// If you are using the [`test_case_map`] macro and [`TestCaseMap`] type, then you should also
/// call this macro to make sure that every entry in a [`TestCaseMap`] has a corresponding `case`
/// statement in [`rstest::rstest`].
///
/// # Examples
///
/// ```ignore
/// test_all_test_cases_ran!(
///     ("test_function_a", &FUNCTION_A_TEST_CASES),
///     ("test_function_b", &FUNCTION_B_TEST_CASES),
/// );
/// ```
macro_rules! test_all_test_cases_ran {
    ($($test_function_test_cases:expr),*) => {
        #[test]
        fn test_all_test_cases_ran() {
            let mut _expected_test_cases = ::std::collections::HashMap::<&str, Vec<&str>>::new();
            $({
                let (test_function_name, test_case_map) = $test_function_test_cases;
                _expected_test_cases
                    .insert(test_function_name, test_case_map.keys().map(|k| *k).collect());
            })*
            $crate::testutil::assert_all_test_cases_ran(&_expected_test_cases, module_path!());
        }
    };
}

pub(crate) use test_all_test_cases_ran;

static GET_TEST_CASE_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r".+::case_\d*_([^:]+)$").unwrap());

static GET_CRATE_NAME_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^[^:]+").unwrap());

static PARSE_TEST_LIST_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^(.+)::case_\d*_(.+): test$").unwrap());

/// Asserts that all [`rstest`] test cases in the given tables of test cases are actual test cases
/// in this test binary.
///
/// The function is not generally to be used directly.  Use the [`test_all_test_cases_ran`] macro
/// instead.
pub(crate) fn assert_all_test_cases_ran(
    expected_test_cases: &HashMap<&str, Vec<&str>>,
    test_module_path: &str,
) {
    // final result for each set are paths like:
    // my_crate_name::my_module::tests::test_frob::success_if_foo

    // test cases built from the user's tast table HashMap
    let expected_test_cases = HashSet::<String>::from_iter(expected_test_cases.iter().flat_map(
        |(test_function_name, test_cases_for_function)| {
            test_cases_for_function.iter().map(move |test_case_name| {
                format!("{test_module_path}::{test_function_name}::{test_case_name}")
            })
        },
    ));

    let crate_name = GET_CRATE_NAME_RE.find(test_module_path).unwrap().as_str();

    // run the test executable to list all tests
    // the output of this command looks like:
    // my_module::tests::test_frob::case_02_success_if_foo: test
    let list_all_tests_output = std::process::Command::new(std::env::current_exe().unwrap())
        .args(["--list", "--format", "terse"])
        .output()
        .unwrap();
    assert_that!(
        list_all_tests_output.status.success(),
        eq(true),
        "Failed to run test EXE to list tests"
    );
    let actual_test_cases = HashSet::<String>::from_iter(
        str::from_utf8(&list_all_tests_output.stdout).unwrap().lines().flat_map(
            |test_output_line| {
                let parsed = PARSE_TEST_LIST_RE.captures(test_output_line)?;
                Some(format!("{crate_name}::{}::{}", &parsed[1], &parsed[2]))
            },
        ),
    );

    let unrun_test_cases: Vec<_> = expected_test_cases.difference(&actual_test_cases).collect();
    assert_that!(
        unrun_test_cases,
        empty(),
        "These test cases are not referenced by the test function with a #[case] attribute."
    );
}

mod tests {
    use googletest::expect_that;

    use super::*;

    #[googletest::test]
    fn test_from_hex() {
        expect_that!(from_hex("AA BB CC"), eq([0xAA, 0xBB, 0xCC]));
    }
}
