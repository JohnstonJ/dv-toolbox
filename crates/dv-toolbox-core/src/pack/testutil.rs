use std::sync::LazyLock;

use display_error_chain::ErrorChainExt;
use googletest::prelude::*;
use num::rational::Ratio;

use super::*;
use crate::{
    file::{Info, UnvalidatedInfo},
    testutil::from_hex,
};

pub(crate) static NTSC: LazyLock<PackContext> = LazyLock::new(|| PackContext {
    file_info: UnvalidatedInfo::new(Info {
        file_size: 120_000,
        video_frame_rate: Ratio::<u32>::new(30_000, 1_001),
        video_duration: Ratio::<u128>::new(1_001, 30_000),
        audio_stereo_stream_count: 1,
        audio_sample_rate: Some(48_000),
    })
    .validate()
    .unwrap(),
});

pub(crate) static PAL: LazyLock<PackContext> = LazyLock::new(|| PackContext {
    file_info: UnvalidatedInfo::new(Info {
        file_size: 144_000,
        video_frame_rate: Ratio::<u32>::from(25),
        video_duration: Ratio::<u128>::new(1, 25),
        audio_stereo_stream_count: 1,
        audio_sample_rate: Some(48_000),
    })
    .validate()
    .unwrap(),
});

/// Shorthand function useful for constructing a validated pack from an unvalidated pack literal.
pub(crate) fn validated<T: PackData>(unvalidated_pack: T, ctx: PackContext) -> ValidPack<T> {
    Unvalidated::new(unvalidated_pack).validate_with(&ctx).unwrap().into()
}

#[derive(Debug)]
pub(crate) struct PackBinaryTestCase<'a> {
    pub(crate) input: &'a str,
    pub(crate) parsed: Option<Pack>,
    pub(crate) err: Option<&'a str>,
    pub(crate) output: Option<&'a str>,
    pub(crate) ctx: PackContext,
}

impl<'a> Default for PackBinaryTestCase<'a> {
    fn default() -> Self {
        Self { input: "", parsed: None, err: None, output: None, ctx: *NTSC }
    }
}

/// Test round trip of a pack: deserialize from binary, check the result, and then back to binary.
pub(crate) fn run_pack_binary_test_case(tc: &PackBinaryTestCase) {
    // Parse input hex to bytes
    let input = from_hex(tc.input);

    // Deserialize the pack and check for expected error and deserialized pack contents
    let (deserialized, err) = Pack::from_raw(&input, &tc.ctx);
    match tc.err {
        None => expect_that!(err, none()),
        Some(msg) => expect_that!(err.map(|e| e.chain().to_string()), some(eq(msg))),
    };
    let expected_pack = match tc.parsed {
        // Leaving tc.parsed as None is a shortcut that avoids retyping the Invalid pack.
        None => Pack::Invalid(
            Type::from(input[0]),
            ValidPack::<Unparsed>::try_from_raw(input[1..].try_into().unwrap(), &NTSC).unwrap(),
        ),
        Some(p) => p,
    };
    expect_that!(deserialized, eq(expected_pack));

    // Serialize the pack and check for expected output
    let serialized = deserialized.to_raw(&tc.ctx);
    // Leaving tc.output as None is a shortcut for saying that the input is expected
    let expected_output = tc.output.map_or_else(|| input, from_hex);
    expect_that!(serialized, eq(expected_output));
}
