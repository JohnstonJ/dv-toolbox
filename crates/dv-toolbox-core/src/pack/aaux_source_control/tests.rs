use arbitrary_int::{u1, u7};
use insta::{assert_debug_snapshot, with_settings};
use itertools::Itertools;
use num::rational::Ratio;
use rstest::rstest;
use stdext::function_name;
use testutil::*;

use super::super::*;
use crate::{pack::testutil::PackBinaryTestCase, testutil::*};

test_all_test_cases_ran!(
    ("test_aaux_source_control_binary", &AAUX_SOURCE_CONTROL_BINARY_TEST_CASES,),
    ("test_aaux_source_control_validation", &AAUX_SOURCE_CONTROL_VALIDATION_TEST_CASES)
);

// ==================== BINARY SERIALIZATION TESTING ====================
// Tests to/from actual/raw DV pack data.

static AAUX_SOURCE_CONTROL_BINARY_TEST_CASES: LazyTestCases<PackBinaryTestCase> = test_case_map!(
    //
    // ===== SOME BASIC SUCCESS CASES =====
    // These are from real tape transfers
    //
    // basic success case: from my Sony DCR-TRV460, first audio channel block
    "basic_success_first_audio_block",
    PackBinaryTestCase {
        input: "51 03 CF A0 FF",
        parsed: Some(Pack::AAUXSourceControl(validated(
            AAUXSourceControl {
                copy_protection: CopyProtection::NoRestriction,
                source_situation: None,
                input_source: Some(InputSource::Analog),
                compression_count: Some(CompressionCount::Compressed1),
                recording_start_point: false,
                recording_end_point: false,
                recording_mode: AAUXRecordingMode::Original,
                insert_channel: None,
                genre_category: None,
                direction: Direction::Forward,
                playback_speed: Some(Ratio::<u8>::ONE),
                reserved: u1::new(0x1),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // basic success case: from my Sony DCR-TRV460, second (empty) audio channel block
    "basic_success_second_empty_audio_block",
    PackBinaryTestCase {
        input: "51 03 FF A0 FF",
        parsed: Some(Pack::AAUXSourceControl(validated(
            AAUXSourceControl {
                copy_protection: CopyProtection::NoRestriction,
                source_situation: None,
                input_source: Some(InputSource::Analog),
                compression_count: Some(CompressionCount::Compressed1),
                recording_start_point: false,
                recording_end_point: false,
                recording_mode: AAUXRecordingMode::InvalidRecording,
                insert_channel: None,
                genre_category: None,
                direction: Direction::Forward,
                playback_speed: Some(Ratio::<u8>::ONE),
                reserved: u1::new(0x1),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // ===== ADDITIONAL CONTRIVED/SYNTHETIC TEST CASES: TEST PLAYBACK SPEEDS =====
    //
    // stopped playback speed
    "stopped_playback_speed",
    PackBinaryTestCase {
        input: "51 03 CF 80 FF",
        parsed: Some(Pack::AAUXSourceControl(validated(
            AAUXSourceControl {
                copy_protection: CopyProtection::NoRestriction,
                source_situation: None,
                input_source: Some(InputSource::Analog),
                compression_count: Some(CompressionCount::Compressed1),
                recording_start_point: false,
                recording_end_point: false,
                recording_mode: AAUXRecordingMode::Original,
                insert_channel: None,
                genre_category: None,
                direction: Direction::Forward,
                playback_speed: Some(Ratio::<u8>::ZERO),
                reserved: u1::new(0x1),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // super slow < 1/16 playback speed
    "super_slow_1_16_playback_speed",
    PackBinaryTestCase {
        input: "51 03 CF 81 FF",
        parsed: Some(Pack::AAUXSourceControl(validated(
            AAUXSourceControl {
                copy_protection: CopyProtection::NoRestriction,
                source_situation: None,
                input_source: Some(InputSource::Analog),
                compression_count: Some(CompressionCount::Compressed1),
                recording_start_point: false,
                recording_end_point: false,
                recording_mode: AAUXRecordingMode::Original,
                insert_channel: None,
                genre_category: None,
                direction: Direction::Forward,
                playback_speed: Some(Ratio::<u8>::new(1, 32)),
                reserved: u1::new(0x1),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // 0 + 1/4 playback speed
    "x_0_and_1_4_playback_speed",
    PackBinaryTestCase {
        input: "51 03 CF 8E FF",
        parsed: Some(Pack::AAUXSourceControl(validated(
            AAUXSourceControl {
                copy_protection: CopyProtection::NoRestriction,
                source_situation: None,
                input_source: Some(InputSource::Analog),
                compression_count: Some(CompressionCount::Compressed1),
                recording_start_point: false,
                recording_end_point: false,
                recording_mode: AAUXRecordingMode::Original,
                insert_channel: None,
                genre_category: None,
                direction: Direction::Forward,
                playback_speed: Some(Ratio::<u8>::new(1, 4)),
                reserved: u1::new(0x1),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // 1/2 + 3/32 playback speed
    "x_1_2_and_3_32_playback_speed",
    PackBinaryTestCase {
        input: "51 03 CF 93 FF",
        parsed: Some(Pack::AAUXSourceControl(validated(
            AAUXSourceControl {
                copy_protection: CopyProtection::NoRestriction,
                source_situation: None,
                input_source: Some(InputSource::Analog),
                compression_count: Some(CompressionCount::Compressed1),
                recording_start_point: false,
                recording_end_point: false,
                recording_mode: AAUXRecordingMode::Original,
                insert_channel: None,
                genre_category: None,
                direction: Direction::Forward,
                playback_speed: Some(Ratio::<u8>::new(19, 32)),
                reserved: u1::new(0x1),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // 32 + 28 playback speed
    "x_32_and_28_playback_speed",
    PackBinaryTestCase {
        input: "51 03 CF FE FF",
        parsed: Some(Pack::AAUXSourceControl(validated(
            AAUXSourceControl {
                copy_protection: CopyProtection::NoRestriction,
                source_situation: None,
                input_source: Some(InputSource::Analog),
                compression_count: Some(CompressionCount::Compressed1),
                recording_start_point: false,
                recording_end_point: false,
                recording_mode: AAUXRecordingMode::Original,
                insert_channel: None,
                genre_category: None,
                direction: Direction::Forward,
                playback_speed: Some(Ratio::<u8>::from(60)),
                reserved: u1::new(0x1),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // unknown playback speed
    "unknown_playback_speed",
    PackBinaryTestCase {
        input: "51 03 CF FF FF",
        parsed: Some(Pack::AAUXSourceControl(validated(
            AAUXSourceControl {
                copy_protection: CopyProtection::NoRestriction,
                source_situation: None,
                input_source: Some(InputSource::Analog),
                compression_count: Some(CompressionCount::Compressed1),
                recording_start_point: false,
                recording_end_point: false,
                recording_mode: AAUXRecordingMode::Original,
                insert_channel: None,
                genre_category: None,
                direction: Direction::Forward,
                playback_speed: None,
                reserved: u1::new(0x1),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // ===== ADDITIONAL CONTRIVED/SYNTHETIC TEST CASES: OTHER TESTS =====
    //
    // various values (1)
    "various_values_1",
    PackBinaryTestCase {
        input: "51 0A 8D 20 7F",
        parsed: Some(Pack::AAUXSourceControl(validated(
            AAUXSourceControl {
                copy_protection: CopyProtection::NoRestriction,
                source_situation: Some(SourceSituation::SourceWithAudienceRestrictions),
                input_source: Some(InputSource::Analog),
                compression_count: Some(CompressionCount::Compressed3OrMore),
                recording_start_point: false,
                recording_end_point: true,
                recording_mode: AAUXRecordingMode::Original,
                insert_channel: Some(AAUXInsertChannel::Channels3_4),
                genre_category: None,
                direction: Direction::Reverse,
                playback_speed: Some(Ratio::<u8>::ONE),
                reserved: u1::new(0x0),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // various values (2)
    "various_values_2",
    PackBinaryTestCase {
        input: "51 93 6F C3 AA",
        parsed: Some(Pack::AAUXSourceControl(validated(
            AAUXSourceControl {
                copy_protection: CopyProtection::OneGenerationOnly,
                source_situation: None,
                input_source: Some(InputSource::Digital),
                compression_count: Some(CompressionCount::Compressed1),
                recording_start_point: true,
                recording_end_point: false,
                recording_mode: AAUXRecordingMode::TwoChannelInsert,
                insert_channel: None,
                genre_category: Some(u7::new(0x2A)),
                direction: Direction::Forward,
                playback_speed: Some(Ratio::<u8>::from(4) + Ratio::<u8>::new(3, 4)),
                reserved: u1::new(0x1),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // all bits set
    "all_bits_set",
    PackBinaryTestCase {
        input: "51 FF FF FF FF",
        parsed: Some(Pack::AAUXSourceControl(validated(
            AAUXSourceControl {
                copy_protection: CopyProtection::NotPermitted,
                source_situation: None,
                input_source: None,
                compression_count: None,
                recording_start_point: false,
                recording_end_point: false,
                recording_mode: AAUXRecordingMode::InvalidRecording,
                insert_channel: None,
                genre_category: None,
                direction: Direction::Forward,
                playback_speed: None,
                reserved: u1::new(0x1),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // all bits clear
    "all_bits_clear",
    PackBinaryTestCase {
        input: "51 00 00 00 00",
        parsed: Some(Pack::AAUXSourceControl(validated(
            AAUXSourceControl {
                copy_protection: CopyProtection::NoRestriction,
                source_situation: Some(SourceSituation::ScrambledSourceWithAudienceRestrictions),
                input_source: Some(InputSource::Analog),
                compression_count: Some(CompressionCount::Compressed1),
                recording_start_point: true,
                recording_end_point: true,
                recording_mode: AAUXRecordingMode::Reserved0,
                insert_channel: Some(AAUXInsertChannel::Channel1),
                genre_category: Some(u7::new(0x00)),
                direction: Direction::Reverse,
                playback_speed: Some(Ratio::<u8>::ZERO),
                reserved: u1::new(0x0),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    }
);

#[googletest::test]
#[rstest]
#[case::basic_success_first_audio_block(function_name!())]
#[case::basic_success_second_empty_audio_block(function_name!())]
#[case::stopped_playback_speed(function_name!())]
#[case::super_slow_1_16_playback_speed(function_name!())]
#[case::x_0_and_1_4_playback_speed(function_name!())]
#[case::x_1_2_and_3_32_playback_speed(function_name!())]
#[case::x_32_and_28_playback_speed(function_name!())]
#[case::unknown_playback_speed(function_name!())]
#[case::various_values_1(function_name!())]
#[case::various_values_2(function_name!())]
#[case::all_bits_set(function_name!())]
#[case::all_bits_clear(function_name!())]
fn test_aaux_source_control_binary(#[case] test_function_name: &str) {
    let tc = AAUX_SOURCE_CONTROL_BINARY_TEST_CASES.get_test_case(test_function_name);
    run_pack_binary_test_case(tc);
}

// ==================== VALIDATION TESTING ====================
// Tests on validation code that was not already tested as part of binary serialization.

static AAUX_SOURCE_CONTROL_VALIDATION_TEST_CASES: LazyTestCases<
    ValidateFailureTestCase<AAUXSourceControl, PackContext>,
> = test_case_map!(
    //
    // invalid genre category
    "invalid_genre_category",
    ValidateFailureTestCase {
        value: AAUXSourceControl {
            copy_protection: CopyProtection::NoRestriction,
            source_situation: Some(SourceSituation::ScrambledSourceWithAudienceRestrictions),
            input_source: Some(InputSource::Analog),
            compression_count: Some(CompressionCount::Compressed1),
            recording_start_point: true,
            recording_end_point: true,
            recording_mode: AAUXRecordingMode::Reserved0,
            insert_channel: Some(AAUXInsertChannel::Channel1),
            genre_category: Some(u7::new(0x7F)), // should be None instead
            direction: Direction::Reverse,
            playback_speed: Some(Ratio::<u8>::ZERO),
            reserved: u1::new(0x0),
        },
        err: "genre_category: instead of specifying Some(0x7F), use None to indicate \
            no information\n",
        ctx: *NTSC
    },
    //
    // invalid playback speed
    "invalid_playback_speed",
    ValidateFailureTestCase {
        value: AAUXSourceControl {
            copy_protection: CopyProtection::NoRestriction,
            source_situation: Some(SourceSituation::ScrambledSourceWithAudienceRestrictions),
            input_source: Some(InputSource::Analog),
            compression_count: Some(CompressionCount::Compressed1),
            recording_start_point: true,
            recording_end_point: true,
            recording_mode: AAUXRecordingMode::Reserved0,
            insert_channel: Some(AAUXInsertChannel::Channel1),
            genre_category: None,
            direction: Direction::Reverse,
            playback_speed: Some(Ratio::<u8>::new(99, 100)), // no way to represent this in binary
            reserved: u1::new(0x0),
        },
        err: "playback_speed: playback speed 99/100 not supported: only playback speeds \
            returned by the valid_playback_speeds function are supported\n",
        ctx: *NTSC
    }
);

#[googletest::test]
#[rstest]
#[case::invalid_genre_category(function_name!())]
#[case::invalid_playback_speed(function_name!())]
fn test_aaux_source_control_validation(#[case] test_function_name: &str) {
    let tc = AAUX_SOURCE_CONTROL_VALIDATION_TEST_CASES.get_test_case(test_function_name);
    run_validate_failure_test_case(tc);
}

// ==================== VALID PLAYBACK SPEEDS SNAPSHOT TESTING ====================
// Tests that the valid playback speeds match a snapshot, which should be compared vs the table in
// the IEC specification.

#[googletest::test]
fn test_valid_playback_speeds() {
    let displayed_speeds: Vec<String> = AAUXSourceControl::valid_playback_speeds()
        .iter()
        .sorted()
        .map(|s| format!("{} {}", s.trunc(), s.fract()))
        .collect();
    with_settings!({prepend_module_to_snapshot => false}, {
        assert_debug_snapshot!(displayed_speeds);
    });
}
