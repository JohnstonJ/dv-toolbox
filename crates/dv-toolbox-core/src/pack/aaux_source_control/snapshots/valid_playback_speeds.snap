---
source: crates/dv-toolbox-core/src/pack/aaux_source_control/tests.rs
assertion_line: 422
expression: displayed_speeds
---
[
    "0 0",
    "0 1/32",
    "0 1/16",
    "0 1/15",
    "0 1/14",
    "0 1/13",
    "0 1/12",
    "0 1/11",
    "0 1/10",
    "0 1/9",
    "0 1/8",
    "0 1/7",
    "0 1/6",
    "0 1/5",
    "0 1/4",
    "0 1/3",
    "0 1/2",
    "0 17/32",
    "0 9/16",
    "0 19/32",
    "0 5/8",
    "0 21/32",
    "0 11/16",
    "0 23/32",
    "0 3/4",
    "0 25/32",
    "0 13/16",
    "0 27/32",
    "0 7/8",
    "0 29/32",
    "0 15/16",
    "0 31/32",
    "1 0",
    "1 1/16",
    "1 1/8",
    "1 3/16",
    "1 1/4",
    "1 5/16",
    "1 3/8",
    "1 7/16",
    "1 1/2",
    "1 9/16",
    "1 5/8",
    "1 11/16",
    "1 3/4",
    "1 13/16",
    "1 7/8",
    "1 15/16",
    "2 0",
    "2 1/8",
    "2 1/4",
    "2 3/8",
    "2 1/2",
    "2 5/8",
    "2 3/4",
    "2 7/8",
    "3 0",
    "3 1/8",
    "3 1/4",
    "3 3/8",
    "3 1/2",
    "3 5/8",
    "3 3/4",
    "3 7/8",
    "4 0",
    "4 1/4",
    "4 1/2",
    "4 3/4",
    "5 0",
    "5 1/4",
    "5 1/2",
    "5 3/4",
    "6 0",
    "6 1/4",
    "6 1/2",
    "6 3/4",
    "7 0",
    "7 1/4",
    "7 1/2",
    "7 3/4",
    "8 0",
    "8 1/2",
    "9 0",
    "9 1/2",
    "10 0",
    "10 1/2",
    "11 0",
    "11 1/2",
    "12 0",
    "12 1/2",
    "13 0",
    "13 1/2",
    "14 0",
    "14 1/2",
    "15 0",
    "15 1/2",
    "16 0",
    "17 0",
    "18 0",
    "19 0",
    "20 0",
    "21 0",
    "22 0",
    "23 0",
    "24 0",
    "25 0",
    "26 0",
    "27 0",
    "28 0",
    "29 0",
    "30 0",
    "31 0",
    "32 0",
    "34 0",
    "36 0",
    "38 0",
    "40 0",
    "42 0",
    "44 0",
    "46 0",
    "48 0",
    "50 0",
    "52 0",
    "54 0",
    "56 0",
    "58 0",
    "60 0",
]
